torch/csrc/jit/dynamic_dag.h

```diff
@@ -0,0 +1,388 @@
+}
+
+template <typename T>
+void DynamicDAG<T>::checkInvariants() {
+  for (size_t ord = 0; ord < vertices_.size(); ++ord) {
+    const auto& maybe_vertex = vertices_.at(ord);
+    if (!maybe_vertex.has_value()) continue;
+    auto* vertex = maybe_vertex.value();
+
+    JIT_ASSERT(vertex->ord == ord);
+    for (auto* v : vertex->in_edges) {
+      JIT_ASSERT(v->ord < ord);
+    }
+    for (auto* v : vertex->out_edges) {
+      JIT_ASSERT(v->ord > ord);
+    }
+  }
+}
+
+
+template <typename T>
+std::pair<vertex_set<T>,vertex_set<T>> DynamicDAG<T>::removeVertex(Vertex<T>* v) {
+  for (auto* parent : v->in_edges) {
+    parent->out_edges.erase(v);
+  }
+  for (auto* child : v->out_edges) {
+    child->in_edges.erase(v);
+  }
+  auto in_edges = std::move(v->in_edges);
+  auto out_edges = std::move(v->out_edges);
+  vertices_[v->ord] = at::nullopt;
+  delete v;
+  return std::make_pair<vertex_set<T>,vertex_set<T>>(std::move(in_edges), std::move(out_edges));
+}
+
+// This is O(|AR|) which is the same thing as O(ord(consumer) - ord(producer))
+// AR is the "affected region": { v s.t. ord(v) in [ord(producer), ord(consumer)] }
+// consisting of the only verticies that can possibly be moved around due
+// to this edge addition.
+template <typename T>
+void DynamicDAG<T>::addEdge(Vertex<T>* producer, Vertex<T>* consumer) {
+  JIT_ASSERT(producer != consumer);
+  if (producer->out_edges.count(consumer)) return;
+  producer->out_edges.insert(consumer);
+  consumer->in_edges.insert(producer);
+
+  if (producer->ord <= consumer->ord) {
+    // topological ordering is already consistent, no need to update.
+    return;
+  }
+
+  // Search for vertices that are reachable from consumer that have a now incorrect
+  // topological ordering.
+  if (!dfsForward(consumer, producer->ord, deltaF_)) {
+    // dfsForward returns false if there is a path from consumer to producer.
+    throw std::runtime_error("Cycle detected while trying to add edge.");
+  }
+
+  // Search for vertices that can reach producer that have a now incorrect
+  // topological ordering
+  dfsBackward(producer, consumer->ord, deltaB_);
+
+  // Reorder the vertices that are reachable from consumer to occur BEFORE
+  // the vertices that can reach producer.
+  reorder(deltaF_, deltaB_);
+}
+
+// Define the affected region for contractEdge(producer, consumer) as
+// { v s.t. ord(v) in [ord(producer), ord(consumer)] }.
+// These are the only vertices that can possibly be moved around
+// during edge contraction.
+//
+// contractEdge is O(|AR| * |out_edges(producer)|)
+template <typename T>
+bool DynamicDAG<T>::contractEdge(Vertex<T>* producer, Vertex<T>* consumer) {
+  JIT_ASSERT(producer != consumer);
+  if (contractionProducesCycle(producer, consumer)) {
+    return false;
+  }
+
+  removeEdge(producer, consumer);
+
+  // We can either merge producer into consumer or the other way around.
+  // I've chosen to merge producer into consumer for now but this is arbitrary.
```

Comment:
If I understand correctly, this choice determines if it's the `in_edges` or `out_edges` insertion that's free, right? We might simple choose the strategy based on their sizes. We sometimes have nodes with large fan in, or large fan out, but rarely with both.

Reply:
Yes, that is correct. The strategy selection is a great idea.